font/src/directwrite/mod.rs

```diff
@@ -14,32 +14,223 @@
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, FontEntry>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    font_collection: FontCollection,
+    font_fallback: Option<FontFallback>,
 }
 
-impl crate::Rasterize for DirectWriteRasterizer {
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        let font_entry = &self.fonts.get(&glyph.font_key).ok_or(Error::FontNotLoaded)?;
+
+        let result = self.rasterize_font_char(&font_entry.font_face, glyph.c, glyph.size);
+
+        // If the glyph is missing from the font, try to get a fallback font, and if found,
+        // rasterize using that font
+        if let Err(Error::MissingGlyph(_)) = &result {
+            let fallback_result =
+                self.get_fallback_font(&font_entry.font, glyph.c).and_then(|fallback_font| {
+                    let fallback_font_face = fallback_font
+                        .ok_or_else(|| Error::MissingGlyph(glyph.c))?
+                        .create_font_face();
+
+                    self.rasterize_fallback_font_char(
+                        &fallback_font_face,
+                        &font_entry.font_face,
+                        glyph.c,
+                        glyph.size,
+                    )
+                });
+
+            if fallback_result.is_ok() {
+                return fallback_result;
+            }
+        }
 
-    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
-        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+        result
     }
 
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn get_fallback_font(&self, font: &Font, c: char) -> Result<Option<Font>, Error> {
+        let fallback = match self.font_fallback.as_ref() {
+            Some(fallback) => fallback,
+            None => return Ok(None),
+        };
+
+        let locale = get_current_locale();
+        let utf16_codepoints = encode_utf16(c);
+        let length = utf16_codepoints.len() as u32;
+
+        let number_subst = dwrote::NumberSubstitution::new(
+            dwrite::DWRITE_NUMBER_SUBSTITUTION_METHOD_NONE,
+            &locale,
+            false,
+        );
+        let text_analysis_source = dwrote::TextAnalysisSource::from_text_and_number_subst(
+            Box::new(TextAnalysisSource { locale, length }),
+            utf16_codepoints,
+            number_subst,
+        );
+
+        let fallback_font = fallback.map_characters(
+            &text_analysis_source,
+            0,
+            length,
+            &self.font_collection,
+            Some(&font.family_name()),
+            font.weight(),
+            font.style(),
+            font.stretch(),
+        );
+
+        Ok(fallback_font.mapped_font)
+    }
+
+    fn rasterize_font_char(
```

Comment:
Yes; I'm struggling to come up with a better name though. Will keep thinking.

Reply:
I can see if I can come up with some ideas once all the other problems are resolved. Since it's just a style issue getting all the other stuff figured out first probably makes more sense.